src/zeit/content/cp/automatic.py

```diff
@@ -67,116 +54,170 @@ def values(self):
                 if teaser is None:
-                    teaser = self._extract_newest(content)
+                    teaser = pop_filter(content)
                     block.change_layout(
                         zeit.content.cp.layout.get_layout('buttons'))
             else:
-                teaser = self._extract_newest(content)
+                teaser = pop_filter(content)
             if teaser is None:
                 continue
             block.insert(0, teaser)
             result.append(block)
 
         return result
 
-    def _retrieve_content(self):
-        if not self._v_try_to_retrieve_content:
-            return []
-        if self.automatic_type == 'channel':
-            content = self._query_solr(
-                self._build_query(), self.query_order)
-        elif self.automatic_type == 'query':
-            content = self._query_solr(
-                self.raw_query, self.raw_order)
-        elif self.automatic_type == 'centerpage':
-            content = self._query_centerpage()
-        else:
-            # BBB
-            if self.raw_query:
-                content = self._query_solr(
-                    self.raw_query,
-                    zeit.content.cp.interfaces.IArea['raw_order'].default)
-            else:
-                content = self._query_solr(
-                    self._build_query(),
-                    zeit.content.cp.interfaces.IArea['raw_order'].default)
-        seen = self._v_retrieved_content
-        self._v_retrieved_content += len(content)
-        if self._v_retrieved_content > seen:
-            self._v_try_to_retrieve_content = True
-        else:
-            self._v_try_to_retrieve_content = False
-        return content
+    @cachedproperty
+    def _content_query(self):
+        return zope.component.getAdapter(
+            self, zeit.content.cp.interfaces.IContentQuery,
+            name=self.automatic_type)
+
+    def select_modules(self, *interfaces):
+        for module in self.values():
+            if any([x.providedBy(module) for x in interfaces]):
+                yield module
+
+
+def pop_filter(items, predicate=None):
+    """Remove the first object from the list for which predicate returns True;
+    no predicate means no filtering.
+    """
+    for i, item in enumerate(items):
+        if predicate is None or predicate(item):
+            items.pop(i)
+            return item
+
+
+def is_lead_candidate(content):
+    metadata = zeit.cms.content.interfaces.ICommonMetadata(content, None)
+    if metadata is None:
+        return False
+    return metadata.lead_candidate
+
+
+class ContentQuery(grok.Adapter):
+
+    grok.context(zeit.content.cp.interfaces.IRenderedArea)
+    grok.implements(zeit.content.cp.interfaces.IContentQuery)
+    grok.baseclass()
+
+    def __init__(self, context):
+        self.context = context
+
+    def __call__(self):
+        return []
+
+    @property
+    def start(self):
+        return self.context.start
+
+    @property
+    def rows(self):
```

Comment:
Das ist eine irre führende Benamung. Ja, so viele Rows sollen gelesen werden, aber welcher Wert steckt dahinter? Besser etwas in Richtung "so viele Teaser sollen laut Schema gelesen werden".